R/tmb-methods.R

```diff
@@ -429,3 +429,91 @@ h_residuals_response <- function(object) {
   assert_class(object, "mmrm_tmb")
   component(object, "y_vector") - unname(fitted(object))
 }
+
+
+#' @describeIn mmrm_tmb_methods Simulates from an mmrm model object
+#' @importFrom stats simulate
+#' @exportS3Method
+#'
+#' @param object (`mmrm`)\cr an fitted 'mmrm' object
+#'
+#' @return A `data.frame` of dimension [n, m] where n is the number of rows in `newdata`,
+#' and m is the number `nsim` of simulated responses.
+simulate.mmrm_tmb <- function(object, nsim = 1,
+                              newdata = NULL,
+                              method = c("conditional", "marginal"),
+                              ...){
+
+  method <- match.arg(method)
+
+  if(is.null(newdata)){
+    newdata <- object$data
+  }
+
+  # make sure new data has the same levels as original data
+  full_frame <- model.frame(
+    object,
+    data = newdata,
+    include = c("subject_var", "visit_var", "group_var", "response_var"),
+    na.action = "na.pass"
+  )
+  tmb_data <- h_mmrm_tmb_data(
+    object$formula_parts, full_frame,
+    weights = rep(1, nrow(full_frame)),
+    reml = TRUE,
+    singular = "keep",
+    drop_visit_levels = FALSE,
+    allow_na_response = TRUE,
+    drop_levels = FALSE
+  )
+
+  if(method == "conditional"){
+    mu <- h_get_prediction(tmb_data, object$theta_est, object$beta_est, object$beta_vcov)
+    ret <- as.data.frame(h_get_sim_per_subj(mu, tmb_data$n_subjects, nsim))
+  }else if(method == "marginal"){
+    ret <- as.data.frame(
+      sapply(seq_len(nsim), function(x) {
+        newtheta <- MASS::mvrnorm(1, object$theta_est, object$theta_vcov)
```

Comment:
if possible, this `newtheta` can be generated in the outer loop. (can be more efficient I think because in mvrnorm there are some expensive computations)

Reply:
As discussed, can create the matrix of all `newtheta` vectors before going into `sapply` or loop to only need one covariance matrix decomposition, hence that will be faster